addons/survey/controllers/main.py

```diff
@@ -302,41 +302,53 @@ def survey_submit(self, survey_token, answer_token, **post):
         # Survey Validation
         access_data = self._get_access_data(survey_token, answer_token, ensure_token=True)
         if access_data['validity_code'] is not True:
-            return {
-                'error': access_data['validity_code'],
-            }
+            return {'error': access_data['validity_code']}
         survey_sudo, answer_sudo = access_data['survey_sudo'], access_data['answer_sudo']
 
+        if answer_sudo.state == 'done':
+            return {'error': _('Unauthorized')}
+
         questions, page_or_question_id = survey_sudo._get_survey_questions(answer=answer_sudo,
                                                                            page_id=post.get('page_id'),
                                                                            question_id=post.get('question_id'))
 
         if not answer_sudo.test_entry and not survey_sudo._has_attempts_left(answer_sudo.partner_id, answer_sudo.email, answer_sudo.invite_token):
             # prevent cheating with users creating multiple 'user_input' before their last attempt
-            return {}
-
-        if not answer_sudo.is_time_limit_reached:
-            # Prepare answers and comment by question
-            prepared_questions = {}
-            for question in questions:
-                answer_full = post.get(str(question.id))
-                answer_without_comment, comment = self._extract_comment_from_answers(question, answer_full)
-                prepared_questions[question.id] = {'answer': answer_without_comment, 'comment': comment}
-
-            # Questions Validation
-            errors = {}
-            for question in questions:
-                answer = prepared_questions[question.id]['answer']
-                comment = prepared_questions[question.id]['comment']
-                errors.update(question.validate_question(answer, comment))
-            if errors:
+            return {'error': _('Unauthorized')}
+
+        # Prepare answers and comment by question
+        prepared_questions = {}
+        for question in questions:
+            answer_full = post.get(str(question.id))
+            answer_without_comment, comment = self._extract_comment_from_answers(question, answer_full)
+            prepared_questions[question.id] = {'answer': answer_without_comment, 'comment': comment}
+
+        if answer_sudo.is_time_limit_reached:
```

Comment:
Commit1: Technically, shouldn't this be located before computing prepared_questions and errors ? That way 1/ we check cheating 2/ we handle answers